cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1674 @@
+// Update buffer profile of a certern PG of a port or all PGs of the port according to its speed, cable_length and mtu
+// Called when
+//    - port's speed, cable_length or mtu updated
+//    - one buffer pg of port's is set to dynamic calculation
+// Args
+//    port - port name
+//    speed, cable_length, mtu - port info
+//    exactly_matched_key - representing a port,pg. when provided, only profile of this key is updated
+// Flow
+// 1. For each PGs in the port
+//    a. skip non-dynamically-calculated or non-exactly-matched PGs
+//    b. allocate/reuse profile according to speed/cable length/mtu
+//    c. check accumulative headroom size, fail if exceeding per-port limit
+//    d. profile reference: remove reference to old profile and add reference to new profile
+//    e. put the old profile to to-be-released profile set
+//    f. update BUFFER_PG database
+// 2. Update port's info: speed, cable length and mtu
+// 3. If any of the PGs is updated, recalculate pool size
+// 4. try release each profile in to-be-released profile set
+task_process_status BufferMgrDynamic::refreshPriorityGroupsForPort(const string &port, const string &speed, const string &cable_length, const string &mtu, const string &exactly_matched_key = "")
+{
+    port_info_t &portInfo = m_portInfoLookup[port];
+    string &gearbox_model = portInfo.gearbox_model;
+    bool isHeadroomUpdated = false;
+    buffer_pg_lookup_t &portPgs = m_portPgLookup[port];
+    set<string> profilesToBeReleased;
+
+    // Iterate all the lossless PGs configured on this port
+    for (auto it = portPgs.begin(); it != portPgs.end(); ++it)
+    {
+        auto &key = it->first;
+        if (!exactly_matched_key.empty() && key != exactly_matched_key)
+        {
+            SWSS_LOG_DEBUG("Update buffer PGs: key %s doesn't match %s, skipped ", key.c_str(), exactly_matched_key.c_str());
+            continue;
+        }
+        auto &portPg = it->second;
+        string newProfile, oldProfile;
+
+        oldProfile = portPg.running_profile_name;
+        if (!oldProfile.empty())
+        {
+            // Clear old profile
+            portPg.running_profile_name = "";
+        }
+
+        if (portPg.dynamic_calculated)
+        {
+            string threshold;
+            //Calculate new headroom size
+            if (portPg.static_configured)
+            {
+                // static_configured but dynamic_calculated means non-default threshold value
+                auto &profile = m_bufferProfileLookup[portPg.configured_profile_name];
+                threshold = profile.threshold;
+            }
+            else
+            {
+                threshold = m_defaultThreshold;
+            }
+            newProfile = getDynamicProfileName(speed, cable_length, mtu, threshold, gearbox_model);
+            auto rc = allocateProfile(speed, cable_length, mtu, threshold, gearbox_model, newProfile);
+            if (task_process_status::task_success != rc)
+                return rc;
```

Comment:
There is a piece of message in [allocateProfile](https://github.com/Azure/sonic-swss/pull/1338/files/80b5b1f1b4dbf7b16415882911cf7253f86c35f5#diff-908ebd83566b58a4ad7eb1133cac45f1fa4f328ada69180a47e496b25a34cee7R512) in case it doesn't return task_success